// preconditions/src/constants.js
exports.ShouldBeDefined = "Variable should be defined.";
exports.ShouldBeUndefined = "Variable should be undefined.";

exports.ShouldBeArray = "Variable should be of type Array.";
exports.ShouldNotBeArray = "Variable should NOT be of type Array.";

exports.ShouldBeObject = "Variable should be of type Object.";
exports.ShouldNotBeObject = "Variable should NOT be of type Object.";

exports.ShouldBeEmpty = "Array or object should be empty.";
exports.ShouldNotBeEmpty = "Array or object should NOT be empty.";

exports.ShouldBeFunction = "Variable should be a Function.";
exports.ShouldNotBeFunction = "Variable should NOT be a Function.";

exports.ShouldBeString = "Variable should be a String.";
exports.ShouldNotBeString = "Variable should NOT be a String.";

exports.ShouldBeNumber = "Variable should be a Number.";
exports.ShouldNotBeNumber = "Variable should NOT be a Number.";

exports.ShouldBeFinite = "Variable should be Finite (i.e. not infinity).";
exports.ShouldBeInfinite = "Variable should be Infinite.";

exports.ShouldBeBoolean = "Variable should be a Boolean.";
exports.ShouldNotBeBoolean = "Variable should NOT be a Boolean.";

exports.ShouldBeDate = "Variable should be a Date.";
exports.ShouldNotBeDate = "Variable should NOT be a Date.";

exports.ShouldBeRegExp = "Variable should be a RegExp.";
exports.ShouldNotBeRegExp = "Variable should NOT be a RegExp.";

exports.ShouldBeFalsey = "Variable should be falsey.";
exports.ShouldNotBeFalsey = "Variable should NOT be falsey.";

exports.IllegalArgument = "Illegal Argument.";
exports.IllegalState = "Illegal State.";

exports.ShouldHaveValidIndex = "Index should be between between 0 (inclusive) and size (exclusive).";
exports.ShouldHaveValidPosition = "Index should be between index between 0 (inclusive) and size (inclusive).";
exports.ShouldHaveValidPositions = "Start and End should be between valid sub range between 0 (inclusive) and size (inclusive).";
exports.StartBeforeEnd = "Start value should be less than the end value.";
